File: tic_tac_toe/ruby/tabuleiro.rb
class Tabuleiro
  TABULEIRO = '
     A   B   C
  1    |   |  
    ---+---+---
  2    |   |  
    ---+---+---
  3    |   |   
  '

  def initialize
    @tabuleiro = TABULEIRO.split "\n"
    @jogador = 'X'
    @matriz = Array.new(3) { |x| Array.new(3) }
  end

  def imprimir
    puts @tabuleiro
    print "Jogador #@jogador, sua Escolha? "
  end

  def preencher(x, y)
    return false if @matriz[x][y]
    @matriz[x][y] = @jogador
    y = y * 2 + 2
    x = x * 4 + 5
    @tabuleiro[y][x] = @jogador
    @jogador = @jogador == 'X' ? "O": "X"
  end

  def jogo_acabou?
    mensagem_final
  end

  def mensagem_final
    @mensagem ||= proc do
      h = horizontais_ou_verticais_iguais
      h ||= diagonais_iguais
      if h
        "#{h} Venceu!"
      elsif todos_preenchidos?
        "Deu velha!"
      end
    end.call
  end

  def horizontais_ou_verticais_iguais
    comp = proc do |m|
      next unless m.all?
      elementos = m.uniq
      return elementos.first if elementos.size == 1
    end

    @matriz.each &comp
    @matriz.transpose.each &comp
    return nil
  end
  private :horizontais_ou_verticais_iguais

  def diagonais_iguais
    return @matriz[0][0] if @matriz[0][0] == @matriz[1][1] && @matriz[1][1] == @matriz[2][2]
    return @matriz[0][2] if @matriz[0][2] == @matriz[1][1] && @matriz[1][1] == @matriz[2][0]
  end
  private :diagonais_iguais

  def todos_preenchidos?
    @matriz.all? { |e| e.all? }
  end
  private :todos_preenchidos?
end

File: tic_tac_toe/scala/Tabuleiro.scala
import scala.collection.mutable.MutableList

class Tabuleiro {
    val TABULEIRO = """
       A   B   C 
    1    |   |   
      ---+---+---
    2    |   |   
      ---+---+---
    3    |   |   
    """
    val tabuleiro = TABULEIRO.split("\n")
    
    val matriz = MutableList[MutableList[String]](
        MutableList(null,null,null), MutableList(null,null,null), MutableList(null,null,null)
    )
    var jogador = "X"
    var m_mensagemFinal = ""

    def imprimir() {
        tabuleiro foreach println
        print("Jogador " + jogador + ", sua escolha? ")
    }

    def preencher(x:Int, y:Int): Boolean = {
        if(matriz(x)(y) != null) return false
        matriz(x)(y) = jogador
        val y1 = y * 2 + 2;
        val x1 = x * 4 + 7;
        val s = tabuleiro(y1)
        tabuleiro(y1) = s.take(x1) + jogador + s.drop(x1 + 1)
        jogador = if(jogador == "X") "O" else "X"
        return true
    }

    def jogoAcabou(): Boolean =  mensagemFinal() != ""

    def mensagemFinal(): String = {
        if(m_mensagemFinal == "") {
            var h = horizontaisOuVerticaisIguais
            if(h == null) h = diagonaisIguais
            if(h != null) {
                m_mensagemFinal = h + " Venceu!"
            } else if(todosPreenchidos) {
                m_mensagemFinal = "Deu velha!"
            }
        }
        return m_mensagemFinal
    }

    private def horizontaisOuVerticaisIguais: String = {
        val comp = (m: MutableList[String]) => {
            if(!m.exists(e => e == null)) {
                val elementos = m.distinct
                if(elementos.size == 1) return elementos.head
            }
        }
        matriz foreach(comp)
        matriz.transpose foreach(comp)
        return null
    }

    private def diagonaisIguais:String = {
        if(matriz(0)(0) == matriz(1)(1) && matriz(1)(1) == matriz(2)(2)) return matriz(0)(0);
        if(matriz(0)(2) == matriz(1)(1) && matriz(1)(1) == matriz(2)(0)) return matriz(0)(2);
        return null;
    }

    private def todosPreenchidos = !matriz.exists(e => e.exists( e2 => e2 == null ) )
}
